### PR TITLE
Fixed Issue #6966 [Status bar overlap on iPads using Tailwind CSS]

### LINKED ISSUE #6966: iOS header overlaps with UI
**Describe the bug**
The iOS header (time, date, battery etc..) overlaps with the UI (see top right of the screen).

**To Reproduce**
Open app

**Expected behavior**
iOS header should not overlap

**Screenshots**
Screenshot 2024-07-09 at 14 15 47 jpeg

**Additional context**
Using the app on iPad (classic)

### PR DESCRIPTION
**Pull request by Xeno-Tek**
# Fix: Status Bar Overlap on iPads

## Overview

This pull request resolves an issue where content overlaps with the status bar on iPad devices. The fix utilizes Tailwind CSS classes to account for the status bar height and safe area insets on iPads without affecting other devices.

## Related Issue

Closes #6966

### `tailwind.config.cjs`

- **Added** iPad-specific breakpoints.
- **Introduced** a new spacing value `ipad-status-bar` for the iPad status bar height.

### `App.tsx`

- **Updated** the main container to include `ipad:pt-ipad-status-bar`, adding proper top padding on iPads.

### `Header.tsx`

- **Adjusted** the header component to account for the status bar on iPads using iPad-specific Tailwind classes.

## Testing

- [x] Verified that the status bar no longer overlaps with content on iPad devices.
- [x] Tested in both portrait and landscape orientations on iPads.
- [x] Ensured the header and main content are properly positioned below the status bar on iPads.
- [x] Used only Tailwind CSS classes for styling adjustments.

### CODE REVIEW

**Comment by Xeno-Tek** [comment]
@VIKTORVAV99, 
I've implemented the changes using Tailwind CSS to address the status bar overlap issue. Please let me know If this works.

**Comment by VIKTORVAV99** [comment]
@Xeno-Tek are those new screenshots?
They still show that the map is not behind the status bar / safe area. Which we wanted to keep.

**Comment by Xeno-Tek** [comment]
@VIKTORVAV99 Thanks for pointing that out! I initially thought it was acceptable for the map not to overlay the safe area insets, but I've gone ahead and adjusted it as per your feedback. The map now correctly overlays the safe area insets. 


1. Updated `main.tsx`:
   * Added CSS variables for safe area insets.
   * Created new utility classes for extending content to the top edge and handling content within safe areas.
   * Added padding to the root element to account for the safe area top inset.
2. Modified `MapWrapper.tsx`:
   * Added the `extend-to-top-edge` class to the map container div to allow it to extend behind the status bar.
3. Updated `ZoomControls.tsx`:
   * Implemented dynamic top margin calculation based on the safe area top inset.
   * Used a `useEffect` hook to update the top margin on component mount and window resize.
   
## Preview Images

### iPad 10th Generation (Portrait)
iPad 10th Generate Fix safe area overlap

### iPad 10th Generation (Landscape)
iPad 10th gen fix horizontaly

### iPhone 16 Pro Max
Iphone 16  Pro Max fix

### iPhone SE 3rd Generation
Iphone SE 3rd gen fix

### iPad Pro 13-inch M4
iPad Pro 13-inch M4

File: mobileapp/ios/App/Podfile [hunk 1]
```diff
@@ -12,7 +12,6 @@ def capacitor_pods
   pod 'Capacitor', :path => '../../node_modules/.pnpm/@capacitor+ios@6.0.0_@capacitor+core@6.0.0/node_modules/@capacitor/ios'
   pod 'CapacitorCordova', :path => '../../node_modules/.pnpm/@capacitor+ios@6.0.0_@capacitor+core@6.0.0/node_modules/@capacitor/ios'
   pod 'CapacitorApp', :path => '../../node_modules/.pnpm/@capacitor+app@6.0.0_@capacitor+core@6.0.0/node_modules/@capacitor/app'
-  pod 'CapacitorShare', :path => '../../node_modules/.pnpm/@capacitor+share@6.0.2_@capacitor+core@6.0.0/node_modules/@capacitor/share'
```
Comment (VIKTORVAV99) [inline mobileapp/ios/App/Podfile:15]: You can't simply remove these, the app will break if you do...

Comment (Xeno-Tek) [reply mobileapp/ios/App/Podfile:15]: My bad for the mix-up, thought they would not get pushed due to git ignore.

File: web/src/main.tsx [hunk 2]
```diff
@@ -18,6 +18,45 @@ import enableErrorsInOverlay from 'utils/errorOverlay';
...
 import { refetchDataOnHourChange } from 'utils/refetching';
 
+const style = document.createElement('style');
+style.textContent = `
+  :root {
+    --sat: env(safe-area-inset-top);
+    --sar: env(safe-area-inset-right);
+    --sab: env(safe-area-inset-bottom);
+    --sal: env(safe-area-inset-left);
+  }
```
Comment (VIKTORVAV99) [inline web/src/main.tsx:28]: are you not simply re-defining existing envs here?

Why do we need this?

File: web/src/main.tsx [hunk 3]
```diff
@@ -18,6 +18,45 @@ import enableErrorsInOverlay from 'utils/errorOverlay';
...
+  .content-within-safe-area {
+    position: relative;
+    z-index: 1;
+    padding: var(--sar) var(--sab) var(--sal);
+  }
+  
+  .pad-for-header {
+    padding-top: var(--sat);
+  }
+`;
```
Comment (VIKTORVAV99) [inline web/src/main.tsx:57]: We already have a index.css file for styles like this. If they are in typescript files they should be tailwind classes, otherwise they should go into the CSS file as these are universal.

File: mockserver/package-lock.json
Comment (VIKTORVAV99) [inline mockserver/package-lock.json:1]: This file should not exist, we use PNPM.

File: mobileapp/ios/App/Podfile.lock [hunk 4]
```diff
@@ -13,6 +13,7 @@ DEPENDENCIES:
   - "CapacitorCordova (from `../../node_modules/.pnpm/@capacitor+ios@6.0.0_@capacitor+core@6.0.0/node_modules/@capacitor/ios`)"
   - "CapacitorShare (from `../../node_modules/.pnpm/@capacitor+share@6.0.2_@capacitor+core@6.0.0/node_modules/@capacitor/share`)"
 
+
```
Comment (VIKTORVAV99) [inline mobileapp/ios/App/Podfile.lock:16]: Avoid diffs in unrelated files please.

File: web/src/features/map-controls/ZoomControls.tsx [hunk 5]
```diff
@@ -1,15 +1,30 @@
...
+          getComputedStyle(document.documentElement).getPropertyValue('--sat'),
+          10
+        ) || 0;
+      setTopMargin(98 + safeAreaTop);
+    };
+
+    updateTopMargin();
+    window.addEventListener('resize', updateTopMargin);
+    return () => window.removeEventListener('resize', updateTopMargin);
+  }, []);
```
Comment (VIKTORVAV99) [inline web/src/features/map-controls/ZoomControls.tsx:20]: This should use a the resizeObserverHook, useEffects should be options of last resort in react. Using a event listener instead of a resizeObserver also have a performance overhead.

File: web/src/features/map-controls/ZoomControls.tsx [hunk 6]
```diff
@@ -1,15 +1,30 @@
...
+
+
   return (
     <NavigationControl
       style={{
         marginRight: 12,
-        marginTop: marginTop,
+        marginTop: topMargin,
```
Comment (VIKTORVAV99) [inline web/src/features/map-controls/ZoomControls.tsx:27]: Why this change?

Please try and keep the diffs to a minimum.

File: web/src/features/map/MapWrapper.tsx [hunk 7]
```diff
@@ -20,7 +20,13 @@ export default function MapWrapper() {
   return (
     <>
       <MapTooltip />
-      {shouldShowFallback ? <MapFallback /> : <Map />}
+      {shouldShowFallback ? (
+        <MapFallback />
+      ) : (
+        <div className="extend-to-top-edge absolute inset-0">
+          <Map />
+        </div>
```
Comment (VIKTORVAV99) [inline web/src/features/map/MapWrapper.tsx:28]: Can this not be handled inside the <Map> component?
We want to avoid creating more elements than is needed.

**Review by VIKTORVAV99 — CHANGES_REQUESTED** [review]
While this seems to work it does this by adding duplicate styles which will be hard to debug later.

There is also files changed in this PR that should not be changed at all which needs to be cleaned up before merging.

**Comment by Xeno-Tek** [comment]
@VIKTORVAV99 Ive made the changes according to your feedback, please check the files changed.


## Test Results After Implementation of Changes

### iPad 10th Gen - Portrait Mode

iPad 10th Gen Portrait Mode

### iPad 10th Gen - Landscape Mode

iPad 10th Gen Landscape Mode

### iPhone 16 Pro Max

iPhone 16 Pro Max

**Comment by VIKTORVAV99** [comment]
Please stop merging master into this. I get a notification every time you do...

**Comment by Xeno-Tek** [comment]
@VIKTORVAV99 @tonypls, could you please let me know when the PR will be reviewed? I'm working on this as part of my assignment, so I won't be able to contribute or make fixes after my report is completed.
